consensus/src/lib.rs

```diff
@@ -1,27 +1,624 @@
+//             return true;
+//         }
+//         // ... then be earliest time received
+//         if pa.time < pb.time {
+//             return false;
+//         }
+//         if pa.time > pb.time {
+//             return true;
+//         }
+//         false
 //     }
 // }
 
+#[allow(dead_code)]
+// TODO: We will generalize it when Lukas will be ready for that. At the moment, he recommended use
+//  types directly.
+// struct Consensus<'a, S: Transactional<'a> + BlockchainStorage> {
+struct Consensus {
+    chain_config: ChainConfig,
+    blockchain_storage: blockchain_storage::Store, //&'a mut S,
+    orphan_blocks: OrphanBlocksPool,
+    highest_block_trust: u64,
+}
+
+type PeerId = u32;
+
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub enum BlockSource {
+    Peer(PeerId),
+    Local,
+}
+
+impl Consensus {
+    // TODO: Make fn that will return tx
+
+    #[allow(dead_code)]
+    pub fn new(chain_config: ChainConfig, blockchain_storage: blockchain_storage::Store) -> Self {
+        Self {
+            chain_config,
+            blockchain_storage,
+            orphan_blocks: OrphanBlocksPool::new_default(),
+            highest_block_trust: 0,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn process_block(
+        &mut self,
+        block: Block,
+        block_source: BlockSource,
+    ) -> Result<Option<BlockIndex>, BlockError> {
+        // self.check_block(&block)?; // Might data will change during the reading from DB? If not, we can move it back here
+        let db_tx = self.blockchain_storage.start_transaction_rw();
+        let mut consensus_ref = ConsensusRef {
+            chain_config: &self.chain_config,
+            db_tx,
+            orphan_blocks: &mut self.orphan_blocks,
+            highest_block_trust: self.highest_block_trust,
+        };
+        consensus_ref.check_block(&block)?;
+        let block_index = consensus_ref.accept_block(&block);
+        if BlockSource::Local == block_source && block_index == Err(BlockError::Orphan) {
+            consensus_ref.new_orphan_block(block)?;
+        }
+        let block_index = block_index?;
+        let result = consensus_ref.activate_best_chain(block_index)?;
+        consensus_ref.commit().expect("Committing transactions to DB failed");
+        Ok(result)
+    }
+}
+
+struct ConsensusRef<'a> {
+    chain_config: &'a ChainConfig,
+    db_tx: <blockchain_storage::Store as Transactional<'a>>::TransactionRw, // TODO: make this generic over Rw and Ro
+    orphan_blocks: &'a mut OrphanBlocksPool,
+    highest_block_trust: u64,
+}
+
+impl<'a> ConsensusRef<'a> {
+    fn commit(self) -> blockchain_storage::Result<()> {
+        self.db_tx.commit()
+    }
+
+    /// Allow to read from storage the previous block and return itself BlockIndex
+    fn get_ancestor(&self, block_id: &Id<Block>) -> Result<BlockIndex, BlockError> {
+        let block_index = self.db_tx.get_block_index(block_id).map_err(BlockError::from)?;
+        match block_index {
```

Comment:
I think the `match` can be removed by using `.ok_or(BlockError::NotFound)?;`

Reply:
Nice catch! 

Let's continue in a new PR: https://github.com/mintlayer/mintlayer-core/pull/124